File: public/code/scalping__range_bounce.py
from common import make_signal, ema, sma, rsi, macd, atr, bollinger_bands, vwap, swing_highs, swing_lows
import math


def _sign(x):
    return (x > 0) - (x < 0)


def rangeBounce(c):
    if len(c) < 35:
        return make_signal(reason="Insufficient data")
    rng = c[-30:]
    rH = max([x['high'] for x in rng])
    rL = min([x['low'] for x in rng])
    size = rH - rL
    a = atr(c, 14)
    ai = len(a) - 1
    # Dar range olmalı (ATR'nin 6 katından az)
    if size / a[ai] > 6:
        return make_signal(reason="Range too wide (trend exists)")
    cur = c[len(c) - 1]
    prev = c[len(c) - 2]
    tol = size * 0.1
    # Alt sınırdan bounce - long
    if cur['low'] <= rL + tol  and  cur['close'] > cur['open']  and  prev['low'] <= rL + tol * 2:
        sl = rL - 0.5 * a[ai]
        r = cur['close'] - sl
        return make_signal(signal="long", entry=cur['close'], stop_loss=sl, take_profit=[(rH + rL) / 2, rH - tol], confidence=0.68, reason=f"Range lower bounce @ {rL}")
    # Üst sınırdan rejection - short
    if cur['high'] >= rH - tol  and  cur['close'] < cur['open']  and  prev['high'] >= rH - tol * 2:
        sl = rH + 0.5 * a[ai]
        r = sl - cur['close']
        return make_signal(signal="short", entry=cur['close'], stop_loss=sl, take_profit=[(rH + rL) / 2, rL + tol], confidence=0.68, reason=f"Range upper rejection @ {rH}")
    return make_signal(reason="Not at range edge")

File: public/code/scalping__range_bounce.ts
import { Candle, Signal, makeSignal, rsi, atr, sma } from "../indicators";

export function rangeBounce(c: Candle[]): Signal {
  if (c.length < 35) return makeSignal({ reason: "Insufficient data" });
  const rng = c.slice(-30);
  const rH = Math.max(...rng.map((x) => x.high)), rL = Math.min(...rng.map((x) => x.low));
  const size = rH - rL;
  const a = atr(c, 14), ai = a.length - 1;
  // Dar range olmalı (ATR'nin 6 katından az)
  if (size / a[ai] > 6) return makeSignal({ reason: "Range too wide (trend exists)" });
  const cur = c[c.length - 1], prev = c[c.length - 2];
  const tol = size * 0.1;
  // Alt sınırdan bounce - long
  if (cur.low <= rL + tol && cur.close > cur.open && prev.low <= rL + tol * 2) {
    const sl = rL - 0.5 * a[ai], r = cur.close - sl;
    return makeSignal({ signal: "long", entry: cur.close, stop_loss: sl, take_profit: [(rH + rL) / 2, rH - tol], confidence: 0.68, reason: `Range lower bounce @ ${rL.toFixed(2)}` });
  }
  // Üst sınırdan rejection - short
  if (cur.high >= rH - tol && cur.close < cur.open && prev.high >= rH - tol * 2) {
    const sl = rH + 0.5 * a[ai], r = sl - cur.close;
    return makeSignal({ signal: "short", entry: cur.close, stop_loss: sl, take_profit: [(rH + rL) / 2, rL + tol], confidence: 0.68, reason: `Range upper rejection @ ${rH.toFixed(2)}` });
  }
  return makeSignal({ reason: "Not at range edge" });
}
